### PR TITLE
gh-109653: Improve the import time of `email.utils`

### PR DESCRIPTION
**Pull request by AlexWaygood**
This patch reduces the import time of `email.utils` by around 46%.

Why do I care about `email.utils`? Well, `email.utils` is imported by `email.message`, and `email.message` is imported by lots of other stdlib modules: `urllib.parse`, `mailbox`, and `importlib.metadata`. Improving the import time of this module has a cascading effect through lots of the rest of the standard library.

A lot of the import cost of `email.utils` has to do with the `make_msgid` function, which requires the `random` and `socket` modules. This function is used by many third-party libraries, but isn't used at all by any of the stdlib modules that import `email.utils`. This patch moves the function into a separate submodule.

* Issue: gh-109653

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2023-09-25-10-47-22.gh-issue-109653.TUHrId.rst
Comment (sobolevn) [inline Misc/NEWS.d/next/Library/2023-09-25-10-47-22.gh-issue-109653.TUHrId.rst]:
```suggestion
Alex Waygood.
```

:)

Comment (AlexWaygood) [reply Misc/NEWS.d/next/Library/2023-09-25-10-47-22.gh-issue-109653.TUHrId.rst]: Ty!!

File: Lib/email/_msgid.py
Comment (AA-Turner) [inline Lib/email/_msgid.py:1]: Minor: ``_utils_msgid`` for explicitness? Non-blocking though, clearly.

File: Lib/email/_msgid.py
Comment (AA-Turner) [inline Lib/email/_msgid.py]: Appreciate this is import time not general refactoring -- but we could consider making this an f-string!
```suggestion
    msgid = f'<{timeval}.{pid}.{randint}{idstring}@{domain}>'
```

File: Lib/email/utils.py
Comment (AA-Turner) [inline Lib/email/utils.py]: The other intra-package imports (at the top-level) use the spelled-out ``email.spam`` form:
```suggestion
        from email._msgid import make_msgid
```

File: Lib/email/utils.py
Comment (AA-Turner) [inline Lib/email/utils.py]: copyedit (I've not been so picky as to insert the definite article, though!):
```suggestion
    # lazy import, to speed up module import time
```

File: Lib/email/_msgid.py
Comment (AlexWaygood) [reply Lib/email/_msgid.py]: I'd rather refrain from (what could be perceived as) gratuitous refactoring here, since I'm not the primary maintainer of this module — I'll defer to the `email` maintainers on what they'd like here 👍

File: Lib/email/_msgid.py
Comment (AlexWaygood) [reply Lib/email/_msgid.py:1]: I'm happy with either; I'll defer to whatever the `email` maintainers prefer :)

File: Lib/email/utils.py
Comment (bitdancer) [inline Lib/email/utils.py]: In general I don't like performance hacks like this, but they do have their place.  I can't speak to whether or not this is a worthwhile performance hack, you should seek approval from the maintainers of the impacted modules for that.

That said, the stdlib itself makes no use of make_msgid, and email.utils is not itself considered part of the new email API.  Moving it into a separate module and making that part of the non-legacy public API of the email module would actually make some sense.  I guess we'd just call it 'msgid'?  Then this code should issue a deprecation warning pointing to the new way to import make_msgid.  It feels kind of weird to have a module with just one function, but there isn't really anything else related to it that I can think of.  (I wonder...maybe make_msgid actually belongs in the UUID module?  Probably not.  Wrong RFC.)

**Comment by hauntsaninja** [comment]
This feels a little strange to me. Should we instead just defer the imports? Might even be faster to do so

**Comment by AlexWaygood** [comment]
> This feels a little strange to me. Should we instead just defer the imports? Might even be faster to do so

Yeah, that might be a better shout, have been meaning to give it a go — just haven't got round to it yet :)

I'm not really enthused about the idea of a painful deprecation period in order to change the place where people are "meant" to import it from. This function is pretty widely used by third-party packages, and the improvement in import time doesn't seem worth the disruption to me :)

**Comment by AlexWaygood** [comment]
I've updated the PR. I abandoned the idea of adding a new submodule; now I just defer the problematic `random`, `socket` and `os` imports to inside `make_msgid()`. The reason I didn't do that originally was because I was concerned it would slow down the `make_msgid()` function, but local experimentation doesn't show that. (I ran `python -m timeit -s "from email.utils import make_msgid" "make_msgid()"` both with and without the change to test.) Shows the importance of benchmarking!

File: Lib/email/utils.py
Comment (AA-Turner) [inline Lib/email/utils.py:184]: Style:

```suggestion
    import socket

```

(I think ``Charset`` can get away without the newline as there's only one line in the pertinent block; here we have a whole function.)

File: Lib/email/utils.py
Comment (hauntsaninja) [inline Lib/email/utils.py:93]: Not sure I like this one, only fine with it since `formataddr` doesn't look too widely used (a lot of the time it's nice to pay these costs upfront, predictable performance is important, e.g. don't want the first request your webserver serves to be randomly slow)

File: Lib/email/utils.py
Comment (hauntsaninja) [inline Lib/email/utils.py]: nit: import os is free

**Review by hauntsaninja — APPROVED** [review]
Looks good

File: Lib/email/utils.py
Comment (AlexWaygood) [reply Lib/email/utils.py]: True, I'll revert that one!

File: Lib/email/utils.py
Comment (AlexWaygood) [reply Lib/email/utils.py:93]: > (a lot of the time it's nice to pay these costs upfront, predictable performance is important

agreed. On the other hand, though, the `email` package goes in quite heavily for lazy imports in some other places, so this does seem in keeping with that general philosophy:

https://github.com/python/cpython/blob/f7860295b16a402621e209871c8eaeeea16f464e/Lib/email/__init__.py#L28-L61

Comment (AlexWaygood) [reply Lib/email/utils.py:93]: I'd be happy to change it so it's imported at the top of the function if you think that'd be better?

Comment (hauntsaninja) [reply Lib/email/utils.py:93]: Nah that'd be worse, module level or here. I'm fine with this as is!

**Pull request merged**
